lib/active_merchant/billing/gateways/ccavenue_mcpg.rb

```diff
@@ -0,0 +1,164 @@
+          post[:billing_city] = options[:billing_address][:city]
+          post[:billing_state] = options[:billing_address][:state]
+          post[:billing_zip] = options[:billing_address][:zip]
+          post[:billing_country] = options[:billing_address][:country]
+		  post[:billing_tel] = options[:billing_address][:phone]
+        end
+
+        if options[:shipping_address]
+          post[:delivery_name] = options[:shipping_address][:name]
+          post[:delivery_address] = options[:shipping_address][:address1]<<", "<<options[:shipping_address][:address2]
+          post[:delivery_city] = options[:shipping_address][:city]
+          post[:delivery_state] = options[:shipping_address][:state]
+          post[:delivery_zip] = options[:shipping_address][:zip]
+          post[:delivery_country] = options[:shipping_address][:country]
+          post[:delivery_tel] = options[:shipping_address][:phone]
+        end
+      end
+
+      def add_invoice(post, money, options)
+		post[:order_id]= options[:order_id]
+        post[:amount] = amount(money)
+        post[:currency] = (options[:currency] || currency(money))
+      end
+
+      def add_payment(post, payment)		
+		post[:card_name] = payment.brand
+        post[:card_number] = payment.number
+        post[:expiry_month] = format(payment.month, :two_digits)
+        post[:expiry_year] = format(payment.year, :four_digits)
+        post[:cvv_number] = payment.verification_value
+      end
+	  
+	  def add_customer_data(post, options)
+		 post[:billing_email] = options[:email] || "email"
+	  end
+
+      def commit(parameters={})
+        url = (test? ? test_url : live_url)
+		key =parameters[:working_key]
+		urlString=post_data(parameters)		
+		enc_response = ssl_post(url,urlString)			
+		response = decrypt(enc_response,key)		
+        Response.new(
+          success_from(response),
+          message_from(response),
+          response,
+          authorization: authorization_from(response),
+          test: test?        
+        )
+      end
+
+      def success_from(response)
+		response["order_status"]=="Success" #true else false will be returned
+      end
+
+      def message_from(response)
+		  result=response["order_status"]
+		  if result=="Success"
+			 return_string ="Your order is Successful"
+	      elsif result=="Failure"
+			return_string ="Your order is Unsuccessful"
+		  elsif result=="Aborted"
+			return_string ="Your transaction is Aborted"
+		  end
+		return return_string
+	  end
+
+      def authorization_from(response) 
+		response["tracking_id"]
+      end
+
+	  INIT_VECTOR = (0..15).to_a.pack("C*") 
+	  
+	  # code for creating encrypted request
+      def post_data(parameters = {})
+		  plain_text = create_plain_text(parameters);
+		  key = parameters[:working_key]
+		  secret_key =  [Digest::MD5.hexdigest(key)].pack("H*") 
+		  cipher = OpenSSL::Cipher::Cipher.new('aes-128-cbc')
+		  cipher.encrypt
+		  cipher.key = secret_key
+		  cipher.iv  = INIT_VECTOR
```

Comment:
@avenuestest do you work for or with CCAvenue? If so, you should let them know that reusing the same IV makes this encryption very vulnerable: https://en.wikipedia.org/wiki/Initialization_vector#SSL_2.0_IV.

On the plus side, it's useless extra work anyhow since the communication is over TLS, so I'll still consider inclusion of CCAvenue in ActiveMerchant.